intermediate_source/dynamo_tutorial.py

```diff
@@ -0,0 +1,491 @@
+for i in range(N_ITERS):
+    inp = generate_data(16)[0]
+    _, dynamo_time = timed(lambda: eval_opt(model, inp))
+    dynamo_times.append(dynamo_time)
+    print(f"dynamo eval time {i}: {dynamo_time}")
+print("~" * 10)
+
+import numpy as np
+eager_med = np.median(eager_times)
+dynamo_med = np.median(dynamo_times)
+speedup = eager_med / dynamo_med
+print(f"(eval) eager median: {eager_med}, dynamo median: {dynamo_med}, speedup: {speedup}x")
+print("~" * 10)
+
+######################################################################
+# And indeed, we can see that running our model with TorchDynamo/TorchInductor
+# results in a significant speedup. On an NVIDIA A100 GPU, we observe a
+# 2x speedup. Speedup mainly comes from reducing Python overhead and
+# GPU read/writes, and so the observed speedup may vary on factors such as model
+# architecture and batch size. For example, if a model's architecture is simple
+# and the amount of data is large, then the bottleneck would be
+# GPU compute and the observed speedup may be less significant.
+#
+# Now, let's consider comparing training.
+
+model = init_model()
+opt = torch.optim.Adam(model.parameters())
+
+def train(mod, data):
+    pred = mod(data[0])
+    loss = torch.nn.CrossEntropyLoss()(pred, data[1])
+    loss.backward()
+
+eager_times = []
+for i in range(N_ITERS):
+    inp = generate_data(16)
+    opt.zero_grad(True)
+    _, eager_time = timed(lambda: train(model, inp))
+    opt.step()
+    eager_times.append(eager_time)
+    print(f"eager train time {i}: {eager_time}")
+print("~" * 10)
+
+model = init_model()
+opt = torch.optim.Adam(model.parameters())
+train_opt = dynamo.optimize("inductor")(train)
+
+dynamo_times = []
+for i in range(N_ITERS):
+    inp = generate_data(16)
+    opt.zero_grad(True)
+    _, dynamo_time = timed(lambda: train_opt(model, inp))
+    opt.step()
+    dynamo_times.append(dynamo_time)
+    print(f"dynamo train time {i}: {dynamo_time}")
+print("~" * 10)
+
+eager_med = np.median(eager_times)
+dynamo_med = np.median(dynamo_times)
+speedup = eager_med / dynamo_med
+print(f"(train) eager median: {eager_med}, dynamo median: {dynamo_med}, speedup: {speedup}x")
+print("~" * 10)
+
+######################################################################
+# Again, we can see that TorchDynamo/TorchInductor takes longer in the first
+# iteration, as it must compile the model, but afterward, we see
+# significant speedups compared to eager. On an NVIDIA A100 GPU, we
+# observe a approximate 2x speedup.
+#
+# One thing to note is that, as of now, we cannot place optimizer code --
+# ``opt.zero_grad`` and ``opt.step`` -- inside of an optimized function.
+# The rest of the training loop -- the forward pass and the backward pass --
+# can be optimized. We are currently working on enabling optimizers to be
+# compatible with TorchDynamo/TorchInductor.
+
+######################################################################
+# Comparison to TorchScript and FX Tracing
+# ------------
```

Comment:
please make the underline on the line 242 as long as the title - it won't fail the build but Sphinx drops warnings for those.